### PR TITLE
re-authentication logic, wip

### PR DESCRIPTION
**Pull request by CSharpRU**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -79,6 +91,170 @@ public function __construct(Transport $transport, LoggerInterface $logger = null
...
+            try {
+                $this->logger->debug('Trying to re-authenticate.');
+
+                if (!$this->authenticate()) {
+                    throw new \RuntimeException('Cannot re-authenticate.');
+                }
+
+                return true;
+            } catch (\Exception $e) {
+                // just skip all exceptions
```
Comment (rumyantseva) [inline src/Client.php]: IMHO, it's important to save attempts count only in the case of failed authentication, so I think `$this->reAuthenticationCounter++;` should be here, not in general function.

File: src/Client.php [hunk 2]
```diff
@@ -79,6 +91,170 @@ public function __construct(Transport $transport, LoggerInterface $logger = null
...
+     * @throws \Vault\Exceptions\ServerException
+     * @throws \RuntimeException
+     * @throws \Psr\Cache\InvalidArgumentException
+     */
+    protected function checkResponse(ResponseInterface $response)
+    {
+        // start re-authentication process if got 403 code and token is expired
+        if (
+            $response->getStatusCode() === 403 &&
+            !$this->isReAuthenticationInProgress() &&
```
Comment (rumyantseva) [inline src/Client.php]: when will `isReAuthenticationInProgress()` return true?

Comment (CSharpRU) [inline src/Client.php]: when reAuthenticationCounter > 0

Comment (rumyantseva) [inline src/Client.php]: obviously :)

but when is it possible? when some request started the reauthentication and some other process was trying to do the same?

Comment (CSharpRU) [inline src/Client.php]: Condition will be always true without it and it'll call reAuthentication every cycle without that `isReAuthenticationInProgress`

**Pull request merged**